### PR TITLE
Team 1 - linear regression helper

### PR DESCRIPTION
**Pull request by alexpondaven**
What's the policy on using extra libraries? Can we use them, or do we want to keep it self-contained.

Was also looking at some libraries for a neural network implementation e.g. https://github.com/patrikeh/go-deep

### CODE REVIEW

**Comment by neeldug** [comment]
> What's the policy on using extra libraries? Can we use them, or do we want to keep it self-contained.
> 
> Was also looking at some libraries for a neural network implementation e.g. https://github.com/patrikeh/go-deep

Extra libraries are fine but do be sure they’re fairly well tested within your means since they introduce other surfaces for bugs to come from. (Also only import the libraries into your own agent’s package).

**Comment by alexpondaven** [comment]
Just the linear regression helpers. Will add q-function integration separately.

File: pkg/.env [hunk 1]
```diff
@@ -0,0 +1,12 @@
+MODE=default
```
Comment (sherwin-dc) [inline pkg/.env:1]: Why is this file added?

Comment (alexpondaven) [reply pkg/.env]: My bad, removed

**Comment by neeldug** [comment]
> I agree with this, I've also moved the other helper function and q-value related files to team1/utils/

I think specifically naming it [`internal`](https://dave.cheney.net/2019/10/06/use-internal-packages-to-reduce-your-public-api-surface), would be good since it quite literally means the API is unexposed outside of the `team1` package.

**Review by neeldug — APPROVED** [review]
Most looks good to me, I think that perhaps some of the maths functions like BoundArray could be implemented in a less copy-paste'd way by means of something like this, and then just passing the array with the function pointer in each place:

```go
func MapSlice[A any](arr []A, f func(A) A) []A {
    arrMapped := make([]A, len(arr))
    for i, v := range arr {
        arrMapped[i] = f(v)
    }
    return arrMapped
}
```

**Comment by alexpondaven** [comment]
> Most looks good to me, I think that perhaps some of the maths functions like `BoundArray` could be implemented in a less copy-paste'd way by means of something like this, and then just passing the array with the function pointer in each place:
> 
> ```go
> func MapSlice[A any](arr []A, f func(A) A) []A {
>     arrMapped := make([]A, len(arr))
>     for i, v := range arr {
>         arrMapped[i] = f(v)
>     }
>     return arrMapped
> }
> ```

Will keep in mind, if there's time I will try to clean those functions up. For now, I also moved the message utils to internal and will merge.
